mllib/src/main/scala/org/apache/spark/ml/reduction/OneVsRestClassifier.scala

```diff
@@ -0,0 +1,175 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.reduction
+
+
+import scala.language.existentials
+
+import org.apache.spark.annotation.{AlphaComponent, DeveloperApi}
+import org.apache.spark.ml.{Estimator, Model}
+import org.apache.spark.ml.attribute.BinaryAttribute
+import org.apache.spark.ml.classification.{ClassificationModel, Classifier, ClassifierParams}
+import org.apache.spark.ml.param.Param
+import org.apache.spark.ml.util.{MetadataUtils, SchemaUtils}
+import org.apache.spark.mllib.linalg._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.{DataFrame, Row}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+import org.apache.spark.storage.StorageLevel
+
+/**
+ * Params for [[OneVsRestClassifier]].
+ */
+private[ml] trait OneVsRestParams extends ClassifierParams {
+
+  type ClassifierType = Classifier[F, E, M] forSome {
+    type F ;
+    type M <: ClassificationModel[F,M];
+    type E <:  Classifier[F, E,M]
+  }
+
+  /**
+   * param for the base classifier that we reduce multiclass classification into.
+   * @group param
+   */
+  val baseClassifier: Param[ClassifierType]  =
+    new Param(this, "baseClassifier", "base binary classifier/regressor ")
+
+  /** @group getParam */
+  def getBaseClassifier: ClassifierType = getOrDefault(baseClassifier)
+
+}
+
+/**
+ *
+ * @param parent
+ * @param baseClassificationModels the binary classification models for reduction.
+ */
+@AlphaComponent
+private[ml] class OneVsRestModel(
+    override val parent: OneVsRestClassifier,
+    val baseClassificationModels: Array[Model[_]])
+  extends Model[OneVsRestModel] with OneVsRestParams {
+
+  /**
+   * Transforms the dataset with provided parameter map as additional parameters.
+   * @param dataset input dataset
+   * @return transformed dataset
+   */
+  override def transform(dataset: DataFrame): DataFrame = {
+    // Check schema
+    val parentSchema = dataset.schema
+    transformSchema(parentSchema, logging = true)
+    val sqlCtx = dataset.sqlContext
+
+    // score each model on every data point and pick the model with highest score
+    val predictions = baseClassificationModels.zipWithIndex.par.map { case (model, index) =>
+      val output = model.transform(dataset)
+      output.select($(rawPredictionCol)).map { case Row(p: Vector) => List((index, p(1))) }
+    }.reduce[RDD[List[(Int, Double)]]] { case (x, y) =>
+      x.zip(y).map { case ((a, b)) =>
```

Comment:
We should leave a TODO here to tune performance. It would be nice if we can use DataFrame expressions in the future to leverage on SQL performance improvements.